sdk/nodejs/provider/server.ts

```diff
@@ -267,6 +269,23 @@ class Server implements grpc.UntypedServiceImplementation {
     }
 
     async constructImpl(call: any, callback: any): Promise<void> {
+        // given that construct calls are serialized, we can attach an uncaught handler to pick up exceptions
+        // in underlying user code. When we catch the error, we need to respond to the gRPC request with the error
+        // to avoid a hang.
+        const uncaughtHandler = (err: Error) => {
+            if (!this.uncaughtErrors.has(err)) {
+                this.uncaughtErrors.add(err);
+                // Use `pulumi.log.error` here to tell the engine there was a fatal error, which should
+                // stop processing subsequent resource operations.
+                log.error(err.stack || err.message || ("" + err));
+            }
+            // bubble the uncaught error in the user code back and terminate the outstanding gRPC request.
+            callback(err, undefined);
```

Comment:
I'm wondering if it should be the following:

```suggestion
            if (!this.uncaughtErrors.has(err)) {
                this.uncaughtErrors.add(err);
                // bubble the uncaught error in the user code back and terminate the outstanding gRPC request.
                callback(err, undefined);
            }
```

I'd added the `log.error` elsewhere, but I think that was misguided -- despite what was originally documented, I don't think it'll stop processing subsequent resource operations. So not sure it does much good here (unless `callback(err, undefined)` doesn't end up showing the error in the `pulumi up` output).

Reply:
Fair enough, it doesn't seem to make a difference in the output. However, the callback should not be made inside the if (this would equate to a gRPC hang in some instances).